### DISCUSSION TITLE
Update the image every time nodes or edges change

### DISCUSSION THREAD

**Discussion by DatTN95** [xyflow/xyflow#4154] [Q&A] [answered]
Base on this example https://reactflow.dev/examples/misc/download-image
It's too lag, i tried useMemo, useCallback but not positive
Please help me!!!
My code
`import { useReactFlow, getNodesBounds, getViewportForBounds } from 'reactflow';
import { toJpeg } from 'html-to-image';
import { useEffect } from 'react';

const imageWidth = 1920;
const imageHeight = 1080;

const ExportImage = (({ nodes, edges, setImgDrawing }) => {
    const { getNodes } = useReactFlow();

    const getDrawing = () => {
        const nodesBounds = getNodesBounds(getNodes());
        const transform = getViewportForBounds(nodesBounds, imageWidth, imageHeight, 0.5, 2);

        toJpeg(document.querySelector('.react-flow__viewport'), {
            backgroundColor: '#ffffff',
            width: imageWidth,
            height: imageHeight,
            style: {
                width: imageWidth,
                height: imageHeight,
                transform: `translate(${transform[0]}px, ${transform[1]}px) scale(${transform[2]})`,
            },
        }).then((dataUrl) => {
            const blob = dataURLToBlob(dataUrl);
            const file = new File([blob], 'Drawing.jpg', { type: blob.type });
            setImgDrawing(file);
        });
    };

    const dataURLToBlob = (dataURL) => {
        const base64 = dataURL.split(',')[1];
        const byteString = atob(base64);
        const arrayBuffer = new ArrayBuffer(byteString.length);
        const uint8Array = new Uint8Array(arrayBuffer);

        for (let i = 0; i < byteString.length; i++) {
            uint8Array[i] = byteString.charCodeAt(i);
        }

        return new Blob([arrayBuffer], { type: 'image/jpeg' });
    };

    useEffect(() => {
        getDrawing();
    }, [nodes, edges]);
});

export default ExportImage;`

**Comment by bcakmakoglu** [comment]
I'm not quite sure why you use this effect?

```ts
useEffect(() => {
    getDrawing();
}, [nodes, edges]);
```

Is there a specific reason you need to draw a new image whenever nodes/edges change? Wouldn't it make more sense to only generate the image when you actually need it (like pressing a download btn or similar)?

**Reply by DatTN95** [reply]
Because i need send image to backend when save flow

**Reply by bcakmakoglu** [reply]
And you save the flow on *any* change? 🤔
What about debouncing it? Seems highly impractical to me to keep saving the flow constantly and sending tons of requests to your BE 😅

**Reply by DatTN95** [reply]
No, it only saves when I press the Save button. If I use the Download button, I need download the image first, and then I need create a function to upload the downloaded image again. So, I need to update the state that contains the image continuously so that it always reflects the latest content image.

**Reply by DatTN95** [reply]
I tried set const { getNodes } = useReactFlow() outside <ReactFlow.../> to handle for save but can't. Do you have suggestion for me?

**Reply by bcakmakoglu** [reply]
So you have a user interaction to save the graph? Why not calculate the image when that button is pressed then?

**Reply by DatTN95** [reply]
i tried like this, but error hook

WI.jsx
`import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch } from 'react-redux';
import { updateBreadcrumb } from '../../Redux/Slices/BreadcrumbSlice';
import { toast } from 'react-toastify';
import DataTable from 'react-data-table-component';
import axios from 'axios';
import { Link, useNavigate, useLoaderData, defer } from 'react-router-dom';
import { format } from 'date-fns-tz';
import { routes } from '../../Configures/Routes';
import { Button, Modal, Card} from 'react-bootstrap';
import { getWIList } from '../../API/WI';
import { useNodesState, useEdgesState } from 'reactflow';
import WorkArea from '../Common/ReactFlow/WorkArea';

const WIList = () => {
    const dispatch = useDispatch();
    const { t } = useTranslation();
    const navigate = useNavigate();
	
	const { getNodes } = useReactFlow();

    const [nodes, setNodes, onNodesChange] = useNodesState([]);
    const [edges, setEdges, onEdgesChange] = useEdgesState([]);
    const [modalWIShow, setModalWIShow] = useState(false);
    const [imgDrawing, setImgDrawing] = useState(null);
	
	 const dataURLToBlob = (dataURL) => {
        const base64 = dataURL.split(',')[1];
        const byteString = atob(base64);
        const arrayBuffer = new ArrayBuffer(byteString.length);
        const uint8Array = new Uint8Array(arrayBuffer);

        for (let i = 0; i < byteString.length; i++) {
            uint8Array[i] = byteString.charCodeAt(i);
        }

        return new Blob([arrayBuffer], { type: 'image/jpeg' });
    };

    const saveWI = async () => {
        try {
		
			let file;
			const nodesBounds = getNodesBounds(getNodes());
			const transform = getViewportForBounds(nodesBounds, imageWidth, imageHeight, 0.5, 2);

			toJpeg(document.querySelector('.react-flow__viewport'), {
				backgroundColor: '#ffffff',
				width: imageWidth,
				height: imageHeight,
				style: {
					width: imageWidth,
					height: imageHeight,
					transform: `translate(${transform[0]}px, ${transform[1]}px) scale(${transform[2]})`,
				},
			}).then((dataUrl) => {
				const blob = dataURLToBlob(dataUrl);
				file = new File([blob], 'Drawing.jpg', { type: blob.type });
			});
		
            const formDataObj = new FormData();
            formDataObj.append('wiInfo', JSON.stringify({ nodes, edges }));
            formDataObj.append('file', file);

            const response = await axios.post(`${import.meta.env.VITE_BACKEND_URL}/WI/UpdateWI12`, formDataObj, {
                headers: {
                    'Content-Type': 'multipart/form-data',
                },
            });

            if (response.data.result) {
                toast.success(`${t('Success')}!`);
                navigate(`${routes.WIWIDetail}?WINumber=${response.data.message}`);
            } else {
                toast.error(response.data.message);
            }
        } catch (error) {
            toast.error(error.message);
        }
    };

    return (
        <>
            <WorkArea nodes={nodes} setNodes={setNodes} onNodesChange={onNodesChange} edges={edges} setEdges={setEdges} onEdgesChange={onEdgesChange} setImgDrawing={setImgDrawing} />
			.....
        </>
    )
}

export { WIList };`

AreaWork.jsxs
`const WorkArea = ({ nodes, setNodes, onNodesChange, edges, setEdges, onEdgesChange, setImgDrawing }) => {
return (
        <>
            <ReactFlow
                deleteKeyCode={'Delete'}
                snapToGrid={snapToGrid}
                nodeTypes={nodeTypes}
                nodes={nodes}
                edges={edges}
                onNodesChange={onNodesChange}
                onEdgesChange={onEdgesChange}
                onNodesDelete={onNodesDelete}
                onConnect={onConnect}
                fitView
            >

                <Panel position='top-right'>
                    <ExportImage setImgDrawing={setImgDrawing} nodes={nodes} edges={edges} />
                    <div className='gap-2 p-2'>
                        <Form.Check type='switch' className='pb-2' id='custom-switch' label={t('SnapToGrid')} onChange={() => setSnapToGrid(!snapToGrid)} checked={snapToGrid} />
                        <div className='form-label fw-bold'>{`${t('Connector')} (${t('SourceTarget')})`}:</div>
                        <Form.Select onChange={(e) => setDrpPosition(e.target.value)} value={drpPosition} className='mb-3'>
                            {dataPosition.map((pos, index) => (
                                <option key={index} value={pos.value}>{pos.label}</option>
                            ))}
                        </Form.Select>

                        <div className='form-label fw-bold'>{t('Object')}:</div>
                        <InputGroup>
                            <InputGroup.Text>{t('Content')}</InputGroup.Text>
                            <Form.Control as='textarea' value={nodeName} onChange={(e) => setNodeName(e.target.value)} onFocus={(e) => e.target.select()} />
                        </InputGroup>

                        <InputGroup>
                            <InputGroup.Text>{t('Degree')}</InputGroup.Text>
                            <Form.Control type='number' placeholder={t('360OfACỉcle')} value={nodeDeg} onChange={(e) => setNodeDeg(e.target.value)} />
                        </InputGroup>

                        <div className='d-flex justify-content-start'>
                            <div style={{ marginRight: '5px' }}>
                                <InputGroup>
                                    <InputGroup.Text>{t('FontSizeL')}</InputGroup.Text>
                                    <Form.Control type='number' placeholder={t('FontSizeL')} value={nodeFontSize} onChange={(e) => setNodeFontSize(e.target.value)} />
                                </InputGroup>
                            </div>
                            <div>
                                <i className={`mdi mdi-format-bold fs-4 ${nodeFontWeight === 600 ? 'text-primary' : ''}`} style={{ cursor: 'pointer' }} onClick={() => setNodeFontWeight(nodeFontWeight === 500 ? 600 : 500)}></i>
                                <i className={`mdi mdi-format-italic fs-4 ${nodeFontStyle === 'italic' ? 'text-primary' : ''}`} style={{ cursor: 'pointer' }} onClick={() => setNodeFontStyle(nodeFontStyle === 'normal' ? 'italic' : 'normal')}></i>
                                <i className={`mdi mdi-format-underline fs-4 ${nodeFontDecoration === 'underline' ? 'text-primary' : ''}`} style={{ cursor: 'pointer' }} onClick={() => setNodeFontDecoration(nodeFontDecoration === 'none' ? 'underline' : 'none')}></i>
                            </div>
                        </div>
                        <div className='d-flex justify-content-start'>
                            <div style={{ marginRight: '5px' }}>
                                <input type='color' className='p-0' value={nodeBgColor} onChange={(e) => setNodeBgColor(e.target.value)} />
                            </div>
                            <div>
                                <input type='color' className='p-0' value={nodeBorderColor} onChange={(e) => setNodeBorderColor(e.target.value)} />
                            </div>
                        </div>
                    </div>
                </Panel>

                <Panel>
                    <div className='gap-2 p-2'>
                        <button className='btn btn-primary' onClick={() => addStation('StationFirst', drpPosition)}>{t('StationFirst')}</button>
                        <button className='btn btn-primary' onClick={() => addStation('Station', drpPosition)}>{t('Station')}</button>
                        <button className='btn btn-primary' onClick={() => addStation('StationEnd', drpPosition)}>{t('StationEnd')}</button>
                        <button className='btn btn-primary' onClick={() => addStation('Human')}>{t('Human')}</button>
                        <button className='btn btn-secondary' onClick={() => addStation('Another', drpPosition)}>{t('Another')}</button>
                    </div>
                </Panel>
                <Background />
                <Controls />
                <MiniMap />
            </ReactFlow>
        </>
    )
}

export default WorkArea;`

**Reply by bcakmakoglu** [reply]
What's the error?

**Reply by DatTN95** [reply]
![Screenshot 2024-04-15 190403](https://github.com/xyflow/xyflow/assets/159091627/a41068de-c19c-40b6-a1d9-1907a060f1c9)
this error, bro!

**Reply by bcakmakoglu** [reply]
[You might wanna check the troubleshooting page.](https://reactflow.dev/learn/troubleshooting) 
That error is outlined on there.

**Reply by DatTN95** [reply]
yes i watched it
reason is this line const { getNodes } = useReactFlow(); write outside <ReactFlow .../>
because i created a component another name as ExportImage.jsx to update image every time nodes or edges change

**Reply by bcakmakoglu** [reply, marked as answer]
You need to wrap your component tree in `<ReactFlowProvider>`.

**Reply by DatTN95** [reply]
yes, it working. thank you for your enthusiastic help
